### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by iaz3** [iaz3/wrye-bash#1]
@Utumno

Ok, so what can i do to help now?

I'm not really familiar enough with the codebase(in all it's goryness) to be comfortable doing complex changes, so is there relatively simple stuff that should be done?

I can probably do the exceptions

**Comment by Utumno** [comment]
The exceptions to me _is_ complex stuff - but yeah go ahead and do some if you found easy ones. What I suggested to you is not complex - it's just blind refactoring - give it a whirl. Basically you need to follow what I do in: https://github.com/wrye-bash/wrye-bash/commit/a4e837f6262507828940e339472b9fb10bec1ed0

Thanks for keeping up !

**Comment by iaz3** [comment]
Thats why i specifically mentioned it.

It's complex, but i dont really need to understand the code to do it. Just see what raises what and put the proper exception there.

**Comment by Utumno** [comment]
Well do some ;)
What about the refactoring of mod import links ?
<Another thing you could do that is definitely not complex is go ahead and squash shadowing warns in `game/game/records.py` - rename type to rec_type and format to fmt.>

**Comment by iaz3** [comment]
I'll attempt that, but i'm not really to comfortable with it.(At Least last i checked on it)

**Comment by Utumno** [comment]
Oh you just replace duplicate code in import links by a call to _Mod_Import_Link._import - adding class variables askTitle, csvFile, progressTitle.

**Comment by iaz3** [comment]
I'll try.

**Comment by iaz3** [comment]
About how much is left to do?

**Comment by Utumno** [comment]
More or less all `Mod_***_Import` links apart from those I done in the commit linked :)

**Comment by Sharlikran** [comment]
Iaz3, just wondering what would you like to do?  The refactoring isn't any less important then anything else.  There are a few records that I am unable to properly decode in Skyrim because it would require a custom unpacker to handle the duplicate subrecord signatures.  It has been done for [Fallout](https://github.com/Sharlikran/wrye-bash/blob/feature-sharlikran-valda-fallout/Mopy/bash/game/falloutnv/records.py#L3787) so that could be used as an example for Skyrim.  I only need to define PERK, PACK, RACE, SCEN.  I would like to define LAND using the TES5Edit simple record setup just to capture the FormIDs but that's about it.

As far as the patchers go I'd like to move the sets of records or bash tags and the attributes for the patcher to the appropriate game\constants.py file.  The Race patcher needs to have a set like the [Actor Value](https://github.com/Sharlikran/wrye-bash/blob/feature-sharlikran-valda-fallout/Mopy/bash/patcher/patchers/importers.py#L889) patcher.  Right now the Race patcher seems like it decides what to retain for the Bash Patch using the Bash tag.  There are [constants like this](https://github.com/Sharlikran/wrye-bash/blob/feature-sharlikran-valda-fallout/Mopy/bash/patcher/patchers/races_multitweaks.py#L800) and [if then statements for the Bash tags like this](https://github.com/Sharlikran/wrye-bash/blob/feature-sharlikran-valda-fallout/Mopy/bash/patcher/patchers/races_multitweaks.py#L847) and what I would like it to look like is more like the [Cell Patcher](https://github.com/Sharlikran/wrye-bash/blob/feature-sharlikran-valda-fallout/Mopy/bash/game/skyrim/constants.py#L17211) so that it's easier to set up for all five games since they will have different records, subrecords, and attributes to patch.

Currently I rebase my branches in this order

wrye-bash : dev
wrye-bash : 151-skyrim-patchers
Sharlikran/wrye-bash : dev-skyrim-patcher-loot-update
Sharlikran/wrye-bash : dev-fallout-patcher-update
Sharlikran/wrye-bash : 268-fallout4-support
Sharlikran/wrye-bash : feature-sharlikran-valda-fallout

So if you want to look at an aggregate collection of the code and all the changes feature-sharlikran-valda-fallout is the branch to look at.  However, if you need to just see what changes the branch provides I think gitk will do that but I'm not a Git Guru.

If you would like to work with records and patchers I am happy to explain what I can't do myself due to my extremely limited knowledge of Python.  Then after the changes you make I can do work on what I need to so it covers the differences of each game.  Valda was helping me but I think other things came up and he ran out of time.

**Comment by iaz3** [comment]
I should probably be able to more successfully test now that i got oblivion
been pretty busy lately with life, and now snowstorm
but i'll try and work on whatever

**Comment by iaz3** [comment]
Need responses to continue work on the stuff i'm doing

**Comment by Utumno** [comment]
I am neck deep into core BAIN refactoring - will look into this as soon as I am finished merging my branches. Meanwhile try to understand what's going on - what should happen and why it does not.

**Comment by iaz3** [comment]
Taking a look at #113 and #293, i'm really at a loss.

I dont know how bash handles it, or what code does what, how it all connects to one another, and what expects what in what condition, i dont know wxPython, etc.

I'll probably be more useful in areas that dont require an intricate knowledge of how the code works, especially across multiple files.

I think i'll look through the issues and see what i think i can do, especially ones regarding Code Quality and the 307 milestones.

**Comment by Utumno** [comment]
Forget about 113,it's really complicated even for me. But 293 would be a big help - if you follow links there you can see how I ripped out  a ScreensDetails panel: https://github.com/wrye-bash/wrye-bash/blob/8b7a031dc48ef916455aaaf87ebb0c38c48178a3/Mopy/bash/basher/__init__.py#L3101-L3130

You could try and do that for the people panel it shouldn't be hard.

**Comment by iaz3** [comment]
I'll try, but idk what methods are expected, or what they are expected to do, or where and how they're called

**Comment by Utumno** [comment]
> I'll try, but idk what methods are expected, or what they are expected to do, or where and how they're called

You just defined software engineering :D

**Comment by iaz3** [comment]
More like reverse engineering

**Comment by Utumno** [comment]
I stand corrected !

The basher API is much simpler to reverse engineer however  - once upon a time it was a big tanggled up mess which I had to reverse enginner. You _can't_ imagine what was in there

It is still a WIP - so difficult to describe. Set a breakpoint in BashFrame.Init to follow what's goin on

Basically there are two hierarchies - the UIList and the Panel - the panels are the tabs.

**Comment by iaz3** [comment]
Got it!

**Comment by iaz3** [comment]
I cant help but think that it might be simpler to just rewrite bash from scratch at this point, considering thats basically whats happening anyway, except much harder since we have to work against the current code..

**Comment by Utumno** [comment]
No it's not. You can try if you want of course

Moreover I already did it, what you see it's the rewrite. Please checkout 304.4 and browse its code, or 305
 That being said there are still things to be done - #293 is one of them

I can come up with other stuff you could do - but trying to "reverse engineer" the code is something you can't get away from. So set breakpoints !

**Comment by iaz3** [comment]
Hmm actually..
if i ever finish up work on my [ModReader](https://github.com/iaz3/ModReader) thingy, all it'd really need is a GUI and then it's a new wrye bash.

Then again,

- i have no idea how bash tags work, or how bain works
- the API doesnt work for anything but reading yet
- it still cant read subgroups correctly(IE, CELL)
- there isnt yet a clear entrypoint
- No clear way to interact with the application(Though, since i discovered PyInstaller can create TRUE one file exe's from python, i can create one now since that was holding me back), or how to interact with it for games
- i dont have any game specific handling done yet since i havent finished the API for that and i keep changing internals..
- there isnt any GUI whatsoever, and i'm not particularly good at making them.

Then again, those are (probably?) relatively small issues compared to the ones in bash...


I'll keep trying on the People Panel thing, make more use of the debugger, and if you have anything else i can do let me know. I tried with the detailspanel earlier today locally, but that didnt pan out, so i'll try again.

**Comment by iaz3** [comment]
Ok, so, um,

At this point in time it's unlikely i will do much work regularly or for a consistent amount of time
or much of anything, really. Work will be done sporadically at best

And there is also a not too unlikely chance i won't, erm, live, erm, long enough to provide a significant contribution to the project.

Long story short things are not going well for me right now.

tl;dr: dont count on me to do anything. Work might be done sporadically, at best, and i have a very real fear that i may kill myself. So, yeah. :/
